src/shims/unix/fd.rs

```diff
@@ -180,14 +197,51 @@ impl FileDescriptor {
         RefMut::map(self.0.borrow_mut(), |fd| fd.as_mut())
     }
 
-    pub fn close<'ctx>(self, communicate_allowed: bool) -> InterpResult<'ctx, io::Result<()>> {
+    pub fn close<'tcx>(
+        self,
+        ecx: &mut MiriInterpCx<'tcx>,
+        communicate_allowed: bool,
+    ) -> InterpResult<'tcx, io::Result<()>> {
         // Destroy this `Rc` using `into_inner` so we can call `close` instead of
         // implicitly running the destructor of the file description.
         match Rc::into_inner(self.0) {
-            Some(fd) => RefCell::into_inner(fd).close(communicate_allowed),
+            Some(fd) => RefCell::into_inner(fd).close(ecx, communicate_allowed),
             None => Ok(Ok(())),
         }
     }
+    pub fn downgrade(&self) -> WeakFileDescriptor {
+        WeakFileDescriptor(Rc::downgrade(&self.0))
+    }
+}
+
+// WeakFileDescriptor is used in epoll ready_list and interest_list to avoid strong references,
+// so the file description can be closed properly.
+#[derive(Clone, Debug, Default)]
+pub struct WeakFileDescriptor(Weak<RefCell<Box<dyn FileDescription>>>);
+
+impl WeakFileDescriptor {
+    pub fn upgrade(&self) -> Option<Rc<RefCell<Box<dyn FileDescription>>>> {
+        self.0.upgrade()
+    }
+}
+
+impl PartialOrd for WeakFileDescriptor {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Eq for WeakFileDescriptor {}
+
+impl PartialEq for WeakFileDescriptor {
+    fn eq(&self, other: &Self) -> bool {
+        std::ptr::eq(self.0.as_ptr(), other.0.as_ptr())
+    }
+}
+
+impl Ord for WeakFileDescriptor {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.as_ptr().cmp(&other.0.as_ptr())
```

Comment:
I don't think we can do this: the order of pointers is non-deterministic, so this can make Miri behave differently when it is executed multiple times and allocations are placed in the address space in different ways.